### PR TITLE
Transparent background for tooltips, dropdowns etc.

### PR DESCRIPTION
**Pull request by Ciantic**
Goal is to allow create overlay windows, such as tooltips, dropdowns etc. by simply setting the backround of the window to transparent. It also allowes to do windows without decorations (#427) and would be better way for creating overlay widgets (#1577). Additionally #1249.

https://xi.zulipchat.com/#narrow/stream/147926-druid/topic/Semi-transparent.20windows.20without.20titlebars

For instance transparency is required to make tooltip shaped window:

Or dropdown window with custom shadows:

Or if you want to make your UI go full WinAmp style from 90's:

---

New API for window builder `transparent`:

```rust
    let main_window = WindowDesc::new(build_root_widget())
        .title("Hello World!")
        .transparent(true) // <-- NEW SETTING HERE
        .window_size((400.0, 400.0));
```

To get fully transparent background, hide also the title bar with `show_titlebar(false)`.

---

Changes include a simple example in `examples/transparency.rs`, currently it renders a test widget like this on a transparent background:

---

Note about Windows 10 code changes: This feature is not supposed to change any behavior of non-transparent windows at all, the opaque windows still should use the same code as before.

The method I chose follows this tutorial: ["Windows with C++ : High-Performance Window Layering Using the Windows Composition Engine"](https://docs.microsoft.com/en-us/archive/msdn-magazine/2014/june/windows-with-c-high-performance-window-layering-using-the-windows-composition-engine), it's the most recent, I asked about that in [window-rs repo](https://github.com/microsoft/windows-rs/issues/541). 

Included are Mac support by @rjwittams, and Linux (GTK) support by @JAicewizard. X11 and web support is stubbed.

### CODE REVIEW

File: druid-shell/src/platform/windows/keyboard.rs
Comment (Ciantic) [inline druid-shell/src/platform/windows/keyboard.rs:611]: This was already commited to master, not sure why it appears here again, I might need to do a new pull request to remove this line.

File: druid-shell/src/platform/windows/window.rs
Comment (JAicewizard) [inline druid-shell/src/platform/windows/window.rs:679]: why should it necessarily be transparent to ignore drop shadows? This also depends on the application, if its 100% transparent shadows would look weird but if its only 10% transparent it would still be nice to have shadows.

File: druid/examples/transparency.rs
Comment (JAicewizard) [inline druid/examples/transparency.rs]: I dont know the druid guidelines around this, but I think its preferred to not have nested imports like this?

File: druid/examples/transparency.rs
Comment (JAicewizard) [inline druid/examples/transparency.rs]: put underscores before unused parameters

File: druid/examples/transparency.rs
Comment (JAicewizard) [inline druid/examples/transparency.rs]: just use a string instead of `LocalizedString`

File: druid/src/app.rs
Comment (JAicewizard) [inline druid/src/app.rs:60]: I am not really sure why resizable is an option either, but to me it would make sense for this to not be a option. its either transparent or not. (defaulting to false of course)

**Comment by JAicewizard** [comment]
you should also ad stubs for any unsupported platform

File: druid-shell/src/platform/windows/window.rs
Comment (Ciantic) [reply druid-shell/src/platform/windows/window.rs:679]: Yeah, I think it's just not possible to run `DwmExtendFrameIntoClientArea` twice. To get the transparency it's already executed above with negative margins.

File: druid/examples/transparency.rs
Comment (Ciantic) [reply druid/examples/transparency.rs]: This was patched by Robert.

File: druid-shell/src/platform/windows/window.rs
Comment (JAicewizard) [inline druid-shell/src/platform/windows/window.rs:1309]: not every transparent window is a popup!!

File: druid/examples/transparency.rs
Comment (Ciantic) [reply druid/examples/transparency.rs]: Yeah, I removed that, VSCode does these automatically, thanks!

File: druid/examples/transparency.rs
Comment (Ciantic) [reply druid/examples/transparency.rs]: Changed to String

File: druid-shell/src/platform/windows/window.rs
Comment (Ciantic) [reply druid-shell/src/platform/windows/window.rs]: I removed that, new way of doing transparency doesn't require WS_POPUP, old way required that.

File: druid/src/app.rs
Comment (Ciantic) [reply druid/src/app.rs:60]: Yeah I agree, however I think the reason it and the others are Option, is probably because it's a builder? In builder pattern I think it's usual to have Option so that to not force the defaults in builder.

**Comment by Ciantic** [comment]
This is now ready for review 🚢, only thing I'm annoyed is the amount of commits there, I think we can't squash this as there are contributions from others, and GitHub doesn't allow squashing only some from the older commits?

**Comment by JAicewizard** [comment]
local squash + force push ;)

**Comment by Ciantic** [comment]
I'm rebasing this, currently a total mess.

**Comment by Ciantic** [comment]
Now I have rebased / squashed this

**Comment by Ciantic** [comment]
Note for posterity, we removed two spurious `println!` lines with force push right there. The Github log doesn't make this very clear if we don't do real commits.

**Comment by Ciantic** [comment]
@cmyr I think this is ready for review.

File: druid/src/app.rs
Comment (JAicewizard) [reply druid/src/app.rs:60]: I cant resolve this for some reason, you'd think it makes sense for the reviewer to be able to mark their own comments are resolved right...
Could you mark it as resolved it for me?

**Comment by raphlinus** [comment]
This causes failure on Windows 7, STATUS_DLL_NOT_FOUND in particular. I haven't investigated in detail, but it's most likely the dependency on DirectComposition. Since it's not supported on Windows 7, it needs to be loaded as optional functions. That existed for a long while but was removed in #1191, specifically commit 290f6e4ffa52e153cf27a25d281a0ff9c00b06d2.

**Comment by Ciantic** [comment]
> it needs to be loaded as optional functions

What do you mean by this? Should we make the `transparent()` function optional somehow?

**Comment by Ciantic** [comment]
It's all green again 🟢. I don't have Windows 7 to test, and looks like the CI doesn't test that either, but when you've got time give it a test on Win 7. 

If it doesn't work, I may have to get Win 7 running in vbox, I don't want to bother you to test the compilation after my changes.

**Comment by raphlinus** [comment]
It works as expected - hello works fine, and the "transparency" example isn't transparent but is otherwise functional. You might consider logging a warning in that case. So this is unblocked on me, but still requires someone to review in all the PR's glorious detail.

**Comment by Ciantic** [comment]
I added the warning if user is using Win 7 and tries transparent, and squashed my win 7 fix changes.

File: druid-shell/src/platform/windows/window.rs
Comment (cmyr) [inline druid-shell/src/platform/windows/window.rs:275]: Let's add a FIXME here to use the `Color::TRANSPARENT` const in piet whenever we do the next release there.

File: druid-shell/src/platform/windows/window.rs
Comment (cmyr) [inline druid-shell/src/platform/windows/window.rs:401]: As of 1.50 there is the very helpful [`bool::then`](http://doc.rust-lang.org/1.50.0/std/primitive.bool.html#method.then) for this sort of situation...

File: druid/examples/transparency.rs
Comment (cmyr) [inline druid/examples/transparency.rs:25]: you probably don't need to implement all of `Widget` here; you can maybe just use a `Painter` widget?

**Review by cmyr — APPROVED** [review]
This looks good to me, a few little observations inline but nothing blocking.

I'm curious if this has been tested on multiple monitors, at all; that seems like a possible source of complications. I'm also slightly scared of macOS, and I'll test again on my laptop with a second monitor when I get home later tonight.

Thanks in any case, this should unblock a lot of other work!

**Comment by Ciantic** [comment]
I pushed some changes.

**Comment by cmyr** [comment]
Okay, this looks good to me, thanks again!

**Pull request merged**
